File: 15-05-2024_account _merge.py
#User function Template for python3
class DisjointSet:
    def __init__(self, n):
        self.parent = list(range(n + 1))
        self.size = [1] * (n + 1)

    def find_parent(self, node):
        if self.parent[node] == node:
            return node
        self.parent[node] = self.find_parent(self.parent[node])
        return self.parent[node]

    def union_by_size(self, u, v):
        par_u = self.find_parent(u)
        par_v = self.find_parent(v)
        if par_u == par_v:
            return
        if self.size[par_u] < self.size[par_v]:
            self.parent[par_u] = par_v
            self.size[par_v] += self.size[par_u]
        else:
            self.parent[par_v] = par_u
            self.size[par_u] += self.size[par_v]


class Solution:
    def accountsMerge(self, accounts):
        mp = {}
        n = len(accounts)
        ds = DisjointSet(n)
        par = 0
        for x in accounts:
            for i in range(1, len(x)):
                mail = x[i]
                if mail in mp:
                    ds.union_by_size(par, mp[mail])
                else:
                    mp[mail] = par
            par += 1
        merged_mail = [[] for _ in range(n)]
        for mail, par_node in mp.items():
            par_node = ds.find_parent(par_node)
            merged_mail[par_node].append(mail)
        ans = []
        for i in range(n):
            if not merged_mail[i]:
                continue
            merged_mail[i].sort()
            temp = [accounts[i][0]] + merged_mail[i]
            ans.append(temp)
        return ans

File: 15-05-2024_account _merge.cpp
// User function Template for C++

class DisjointSet{
    vector<int> parent, size;
public:    
    DisjointSet(int n = 0){
        parent.resize(n+1);
        size.resize(n+1, 1);
        for(int i=0; i<=n; ++i){
            parent[i] = i;
        }
    }
    int findParent(int node){
        if(parent[node] == node)return node;
        return parent[node] = findParent(parent[node]);
    }

    void unionBySize(int u, int v){
        int parU = findParent(u);
        int parV = findParent(v);
        if(parU == parV)return;
        if(size[parU] < size[parV]){
            parent[parU] = parV;
            size[parV] += parU;
        }else{
            parent[parV] = parU;
            size[parU] += parV;
        }
    }
};


class Solution{
  public:

    vector<vector<string>> accountsMerge(vector<vector<string>> &accounts) {
        unordered_map<string, int> mp;
        int n = accounts.size();
        DisjointSet ds(n);
        int par = 0;
        for(vector<string> &x : accounts){
            for(int i=1; i<x.size(); ++i){
                string mail = x[i];
                if(mp.find(mail) != mp.end()){
                    ds.unionBySize(par, mp[mail]);
                }else{
                    mp[mail] = par;
                }
            }
            par++;
        }
        vector<string> mergedMail[n];
        for(auto it: mp){
            string mail = it.first;
            int parNode = ds.findParent(it.second);
            mergedMail[parNode].push_back(mail);
        }
        vector<vector<string>> ans;
        for(int i=0; i<n; ++i){
            if(mergedMail[i].size() == 0)continue;
            sort(mergedMail[i].begin(), mergedMail[i].end());
            vector<string> temp;
            temp.push_back(accounts[i][0]);
            for(auto &x : mergedMail[i]){
                temp.push_back(x);
            }
            ans.push_back(temp);
        }
        return ans;
    }
};
